### PR TITLE
osd: Give recovery for inactive PGs a higher priority

### PR DESCRIPTION
**Pull request by dzafman**
New pool priority range of -10 to 10
Adjust tests for new priority values
Add backfill priority log message and remove redundanacy to recovery priority

Fixes: https://tracker.ceph.com/issues/39099

Signed-off-by: David Zafman <dzafman@redhat.com>

- [x] References tracker ticket
- [x] ~Updates documentation if necessary~
- [x] Includes tests for new functionality or reproducer for bug

### CODE REVIEW

**Comment by dzafman** [comment]
@neha-ojha We should talk about this change. The more priorities change the more mixed clusters would see unexpected recovery/backfill ordering.

Possible changes:
1. Could do OSD_RECOVERY_INACTIVE_PRIORITY_BASE = OSD_BACKFILL_INACTIVE_PRIORITY_BASE
1. Could reduce the range for pool priorities to 20 and try to minimize priority changes

File: src/osd/PG.cc [hunk 1]
```diff
@@ -2471,15 +2471,22 @@ inline int PG::clamp_recovery_priority(int priority)
...
   // a higher value -> a higher priority
-  int64_t ret = 0;
+  int ret = OSD_RECOVERY_PRIORITY_BASE;
 
   if (state & PG_STATE_FORCED_RECOVERY) {
     ret = OSD_RECOVERY_PRIORITY_FORCED;
   } else {
-    pool.info.opts.get(pool_opts_t::RECOVERY_PRIORITY, &ret);
-    ret = clamp_recovery_priority(OSD_RECOVERY_PRIORITY_BASE + ret);
+    if (acting.size() < pool.info.min_size) {
```
Comment (dzafman) [inline src/osd/PG.cc]: @neha-ojha This is wrong.  During recovery acting == up.  So if active.size() < pool.info.min_size, the result of the recovery will still be too low to allow I/O.  This would be a very uncommon case plus it doesn't help us to prioritize doing that recovery.  Also, recovery allows simultaneous operations because even if data hasn't transferred, those operations can be logged on all OSDs.  I'm not sure how this interacts with pg log size limiting.  If recovery is slow due to large objects and clients send lots of small operations, the count of log entries is going to go up.  Will log trimming slow client requests, so the log doesn't grow unbounded?

When recovery is degraded, we don't differentiate whether or not there are at least min_size replicas of every object.   In that case we would want a priority boost, but I'm not sure the best way to determine between the 2 cases.  Also, as recovery proceeds the degraded will go down, but at the point that example 1 has 100 degraded like example 2 it is really 100 objects on OSD 1 and 50 objects on OSD 0 and OSD 2. So in actual fact 50 objects are still below min_size assuming we push the same object to both OSDs as recovery goes along.

Example 1, size 3 min_size 2 replicated pool 
Should get a priority boost because for 100 objects has 200 degraded (below min_size with 100 replicas existing)
PG 1.0 [1, 0, 2] active+clean
OSDs 0 and 2 go down and out
PG 1.0 [1, 3, 4].  active+recovering+degraded

Example 2, size 3 min_size 2 replicated pool 
No priority boost because 100 objects has 100 degraded (min_size 2 ok with 200 existing replicas)
PG 1.0 [1, 0, 2] active+clean
OSDs 2 go down and out
PG 1.0 [1, 0, 3].  active+recovering+degraded

File: src/mgr/DaemonServer.cc [hunk 2]
```diff
@@ -1553,14 +1562,15 @@ bool DaemonServer::_handle_command(
...
-	 << "unavailable";
+      ss << dangerous_pgs << " PGs are already too degraded, would become"
+	 << " too degraded or might become unavailable";
       cmdctx->reply(-EBUSY, ss);
       return true;
     }
     ss << "OSD(s) " << osds << " are ok to stop without reducing"
-       << " availability, provided there are no other concurrent failures"
-       << " or interventions. " << pg_delta.size() << " PGs are likely to be"
+       << " availability, risking data, provided there are no other concurrent failures"
```
Comment (liewegas) [inline src/mgr/DaemonServer.cc]: ``... availability or risking data, ...``

File: src/mgr/DaemonServer.cc [hunk 3]
```diff
@@ -1433,7 +1434,6 @@ bool DaemonServer::_handle_command(
       if (!r) {
         ss << "OSD(s) " << osds << " are safe to destroy without reducing data"
            << " durability.";
-        safe_to_destroy.swap(osds);
```
Comment (liewegas) [inline src/mgr/DaemonServer.cc]: why is this line changed?

File: src/mon/PGMap.cc [hunk 4]
```diff
@@ -1305,8 +1305,10 @@ void PGMap::stat_pg_add(const pg_t &pgid, const pg_stat_t &s,
     num_pg_by_osd[*p].acting++;
   }
   for (auto p = s.up.begin(); p != s.up.end(); ++p) {
-    pg_by_osd[*p].insert(pgid);
-    num_pg_by_osd[*p].up++;
+    if (pg_by_osd[*p].find(pgid) == pg_by_osd[*p].end()) {
```
Comment (liewegas) [inline src/mon/PGMap.cc]: maybe use ``auto& t = pg_by_osd[*p];`` to avoid using operator[] 3 times here?

File: src/mgr/DaemonServer.cc (hunk 3 above)
Comment (dzafman) [reply src/mgr/DaemonServer.cc]: @liewegas At this point osds = safe_to_destroy.  I almost added an assert to show that, but decided it as overkill.  I tested it with the assert.

File: src/mon/OSDMonitor.cc [hunk 5]
```diff
@@ -13178,3 +13180,46 @@ void OSDMonitor::_pool_op_reply(MonOpRequestRef op,
...
+    }
+  }
+  if (max_prio == 0) {
+    dout(20) << __func__ << " nothing to fix" << dendl;
+    return;
+  }
+  // Current pool priorities exceeds new maximum
+  for (const auto &i : osdmap.get_pools()) {
+       const auto pool_id = i.first;
+       pg_pool_t pool = i.second;
```
Comment (liewegas) [inline src/mon/OSDMonitor.cc]: indentation?

**Comment by dzafman** [comment]
@liewegas @neha-ojha I've now switched to pool priorities in the range -10 to 10.  Also, the mon converts the pools if it finds them outside this range.  A test verifies that various scenarios scale to 0 to 10, -10 to 0 or -10 to 10 depending on the initial priorities.

**Comment by dzafman** [comment]
jenkins test docs

**Comment by dzafman** [comment]
render docs

**Comment by dzafman** [comment]
jenkins render docs

**Comment by dzafman** [comment]
jenkins doc build

**Comment by dzafman** [comment]
@neha-ojha @liewegas  I've change the conversion so that it leaves unset pool priorities alone as well as negative or positive priorities if they are in range.  For example:

Original pool priorities: -5 0 15 30
Previous version: -10 -7 1 10
After:  -5 0 5 10

**Comment by dzafman** [comment]
Testing passed
dzafman-2019-04-24_12:58:27-rados-wip-zafman-testing2-distro-basic-smithi
Standalone only - 7 passed
dzafman-2019-04-24_09:08:32-rados-wip-zafman-testing-distro-basic-smithi
rados suite only 3 unrelated failures
3888410 DEAD waiting for mon quorum
3888034 ceph pg dump timed out 
3888476 saw valgrind issues

File: doc/rados/operations/pools.rst [hunk 6]
```diff
@@ -602,7 +602,7 @@ You may set values for the following keys:
 ``recovery_priority``
 
 :Description: When a value is set it will boost the computed reservation priority
-              by this amount. This value should be less than 30.
+              relative to this amount. This value must be in the range -10 to 10.
```
Comment (neha-ojha) [inline doc/rados/operations/pools.rst]: we should update `doc/rados/operations/placement-groups.rst` to say that negative values can now be assigned for less important pools.

File: src/mgr/DaemonServer.cc [hunk 7]
```diff
@@ -1514,35 +1514,44 @@ bool DaemonServer::_handle_command(
...
-	for (auto osd : osds) {
-	  auto p = pg_map.pg_by_osd.find(osd);
-	  if (p != pg_map.pg_by_osd.end()) {
-	    for (auto& pgid : p->second) {
-	      --pg_delta[pgid];
+	for (const auto& q : pg_map.pg_stat) {
+          set<int32_t> pg_acting;  // net acting sets (with no missing if degraded)
+	  bool found = false;
+	  if (q.second.state & PG_STATE_DEGRADED) {
+	    //ss << q.first << " avail_no_missing " << q.second.avail_no_missing << " ";
```
Comment (neha-ojha) [inline src/mgr/DaemonServer.cc]: do we want to include this?

File: src/mgr/DaemonServer.cc [hunk 8]
```diff
@@ -1514,35 +1514,44 @@ bool DaemonServer::_handle_command(
...
+	    for (auto& anm : q.second.avail_no_missing) {
+	      if (osds.count(anm.osd)) {
+		found = true;
+		continue;
+	      }
+	      pg_acting.insert(anm.osd);
+	    }
+	  } else {
+	    //ss << q.first << " acting " << q.second.acting << " ";
```
Comment (neha-ojha) [inline src/mgr/DaemonServer.cc]: ditto

**Comment by dzafman** [comment]
@neha-ojha I added 2 commits to squash with your suggested changes.

**Comment by neha-ojha** [comment]
> @neha-ojha I added 2 commits to squash with your suggested changes.

looks good

**Comment by dzafman** [comment]
My rados run which took 2 days finally finished.  A bunch of failures were probably cause by my pushing a new "wip-zafman-testing" so that the tests got git checkout failures:
I'm putting the link here for reference.
http://pulpito.ceph.com/dzafman-2019-04-24_09:08:32-rados-wip-zafman-testing-distro-basic-smithi/
